add retry to functions

Jonahss opened the pull request.

jlipps commented:
rather than creating functions just to wrap with retry, you could retry the execs inside, that way we still just have the one function. or what was your reasoning behind doing it this way?

sebv commented:
Personally I would prefer getVersion()/getVersionWithRetry(), it's not really good idea to hide the magic.

You could write it like this:

```
async function _getVersion() {

}

getVersion = cached(_getVersion)

getVersionWithRetry = retryFn(3, getVersion)
```

```
cached = similar to _.memoize (You don't really need a cache when you have the closure)
retryfn: similar as asyncbox.retry, but operating at function level. 
```

jlipps commented:
oh yeah, underscore/lodash already have a nice way of memoizing function return values so we don't need to make our own cache and store state / etc.

i'm also OK to be more transparent in our naming conventions about `getVersion` vs `getVersionWithRetry`. In that case, as a library user I'd expect `getVersionWithRetry` to take an optional param, num retries. Then we export both. So something like:

``` js
...

async function getVersionWithRetry (tries = 3) {
  return retry(tries, getVersion);
}

...

export { getVersion, getVersionWithRetry };
```

Jonahss commented:
Ok, edited it as per an offline discussion with @jlipps 

I added the 'without retry' functions to the exports, so non-appium users of the repo could make us of them. We still want to make them use the Retry ones though, since that's the best way to have the functions work.

We also added an argument to the retried functions, this should highlight the fact that a retry is happening and not hide it. But if you're unaware you end up using the best solution. (plus users can control the number of retries).

jlipps commented on lib/xcode.js:
```diff
@@ -170,4 +180,8 @@ async function getMaxIOSSDK () {
+  return retry(retries, getMaxIOSSDKWithoutRetry);
+}
+
+export {getPath, getVersion, getAutomationTraceTemplatePath, getMaxIOSSDK, getAutomationTraceTemplatePathWithoutRetry, getMaxIOSSDKWithoutRetry};
```
last comment, please 1. add spaces around export braces, and 2. line up second line with beginning of first export column
